text/0031-transactions-with-contingent-inputs.md

```diff
@@ -0,0 +1,394 @@
+ERC20 representation of locked bitcoin which can be used more easily on the Ethereum network.) Another comment here is that in these multistep protocols,
+if the counterparty stops participating, you can still recover your funds, but you still pay a transaction fees to recover them from the escrow, and you
+didn't manage to complete your swap. You can try again with a different counterparty, but if this happens repeatedly you will lose money to fees.
+If there are a lot of bad actors in the network, then this could happen often, and it may be very difficult for the network to prevent this from happening
+when the underlying mechanism is a multistep protocol.
+
+This proposal is considerably simpler -- the rounds of interaction are a minimum, requiring only one side to send a transaction fragment to another.
+No one is required to "escrow" their funds into a smart contract or temporary address. A party who desires to swap can broadcast their transaction fragment
+to many potential counterparties. Because the Ring MLSAG contains the key image of the spent input, everyone is able to check if the transaction fragment is
+still valid, and the initiating party can cancel the swap proposal by simply spending the input to themselves in a separate transaction, burning its key image
+and preventing any of the broadcasted transaction fragments from being used in a valid transaction.
+
+A feature like this could potentially be used to build a fast, clean asset-swap experience in a mobile app, without relying on smart contracting functionality.
+Note that, much like with current app experiences, there is no "pending" state for a swap, it either happens atomically or it doesn't before the tombstone block,
+and so the app becomes responsible for tracking pending swaps just as with pending transfers.
+
+More generally, it may be interesting to explore ways that users can collaboratively sign transactions securely. This is one way that we can experiment with this,
+and could potentially be extended in the future.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In a MobileCoin transaction, an input consists of a collection of `TxOut`'s in the ledger, one of which is truly being spent and others of which are just being mixed in. This ring,
+and the proofs of membership, appear in the `TxIn`, in the `TxPrefix`. A transaction produces several new `TxOut`'s, and these outputs appear in the `TxPrefix`.
+
+The signature part of a transaction contains several things:
+* A `RingMLSAG` for each `TxIn`
+* A `pseudo_output_commitment` for each `TxIn`
+* A range proof for each `pseudo_output_commitment`, relative to the generator for this token id.
+
+Transaction validation ensures that the sum of outputs minus the sum of pseudo-outputs is the zero point on the elliptic curve, (including the implicit fee output)
+which ensures that the transaction is valid.
```

Comment:
If I recall my Risretto math correctly, there are different prime subgroups which a token might belong to right - do each token types inputs and outputs independently sum to the zero point? (Say if you have MoB and MeowB the MoB-only inputs/outputs sum to the zero point & the MeowB inputs + ouputs independently sum to the zero point?)

Reply:
Yeah, so, suppose G_1 is the base for MOB and G_2 is the base for MEOWB, and H is the blinding factor.

The transaction math checks that the sum of outputs + fee output - sum of pseudo outputs is equal to 0, and actually, the zero point of the elliptic curve:

https://github.com/mobilecoinfoundation/mobilecoin/blob/acbf10842b0d17a22cce0a2123886f03bebea487/transaction/core/src/ring_signature/rct_bulletproofs.rs#L195

So effectively, each output and pseudooutput that we add here is a vector Pedersen commitment, so this is like

(output_1_value * G_1 + blinding * H) + (output_2_value * G_2 + blinding * H) + (fee_value * G_1 + blinding * H) - (input_1_value * G_1 + blinding * H) - (input_2_value * G_2 + blinding * H) = 0

If we collect terms, this is like

(MOB_balance * G_1) + (MEOWB_balance * G_2) + (balance_of_blinding_factors * H) = 0

At this point, we want to use the idea that the generators are all orthogonal -- it is intractable for any person to find a nontrivial linear relation between `G_1, G_2, H`. This is happening because they were hashed to the curve, and decision diffie hellman impllies that random curve points are orthogonal. (also mentioned in comments on MCIP 25)

So, if the person who built the transaction can't find a nontrivial linear relation, then to satisfy the equation they must have made a trivial linear relation. So

* `MOB_balance = 0`
* `MEOWB_balance = 0`
* `balance_of_blinding_factors = 0`

So this is the "vector pedersen commitment" idea (also in bulletproofs), where like, we can basically think of `G_1, G_2, H` as a basis for some vectorspace. So this is why its okay to have mixed inputs and outputs in the balance check part now.

The only reason that the token ids even need to be revealed to the enclave here, is that we have to check the validity of the range proofs, and we can't do that if we don't know the generator for that token.

Reply:
Thanks for explaining this!